File: backend/src/modules/system/system.service.ts
import { Injectable } from '@nestjs/common';
import { KamailioRpcService } from '../../common/services/kamailio-rpc.service';

@Injectable()
export class SystemService {
  constructor(private kamailioRpcService: KamailioRpcService) {}

  // ========== 核心信息 ==========
  async getCoreInfo(): Promise<any> {
    return this.kamailioRpcService.call('core.info');
  }

  async getVersion(): Promise<any> {
    return this.kamailioRpcService.call('core.version');
  }

  async getUptime(): Promise<any> {
    return this.kamailioRpcService.call('core.uptime');
  }

  async getSharedMemory(): Promise<any> {
    return this.kamailioRpcService.call('core.shmmem');
  }

  async getPrivateMemory(): Promise<any> {
    return this.kamailioRpcService.call('core.ppdefines');
  }

  async getProcessList(): Promise<any> {
    return this.kamailioRpcService.call('core.psx');
  }

  // ========== 配置管理 ==========
  async getConfig(group: string, name: string): Promise<any> {
    return this.kamailioRpcService.call('cfg.get', [group, name]);
  }

  async setConfigNow(group: string, name: string, value: any): Promise<void> {
    if (typeof value === 'number') {
      await this.kamailioRpcService.call('cfg.set_now_int', [group, name, value]);
    } else {
      await this.kamailioRpcService.call('cfg.set_now_string', [group, name, value]);
    }
  }

  async listConfig(): Promise<any> {
    return this.kamailioRpcService.call('cfg.list');
  }

  async diffConfig(): Promise<any> {
    return this.kamailioRpcService.call('cfg.diff');
  }

  // ========== TLS ==========
  async getTlsList(): Promise<any> {
    return this.kamailioRpcService.call('tls.list');
  }

  async getTlsInfo(): Promise<any> {
    return this.kamailioRpcService.call('tls.info');
  }

  async reloadTls(): Promise<void> {
    await this.kamailioRpcService.call('tls.reload');
  }

  // ========== Pike (防攻击) ==========
  async getPikeList(): Promise<any> {
    return this.kamailioRpcService.call('pike.list');
  }

  async getPikeTop(limit: number = 10): Promise<any> {
    return this.kamailioRpcService.call('pike.top', [limit]);
  }

  // ========== 统计信息 ==========
  async getStatistics(group?: string): Promise<any> {
    const params = group ? [group] : ['all'];
    return this.kamailioRpcService.call('stats.get_statistics', params);
  }

  async resetStatistics(name: string): Promise<void> {
    await this.kamailioRpcService.call('stats.reset_statistics', [name]);
  }

  async clearStatistics(name: string): Promise<void> {
    await this.kamailioRpcService.call('stats.clear_statistics', [name]);
  }

  // ========== 模块管理 ==========
  async getModulesList(): Promise<any> {
    return this.kamailioRpcService.call('core.modules');
  }

  // ========== 综合状态 ==========
  async getSystemStatus(): Promise<any> {
    const [info, uptime, shmmem, processes] = await Promise.all([
      this.getCoreInfo().catch(() => null),
      this.getUptime().catch(() => null),
      this.getSharedMemory().catch(() => null),
      this.getProcessList().catch(() => []),
    ]);

    return {
      info,
      uptime,
      memory: shmmem,
      processCount: Array.isArray(processes) ? processes.length : 0,
    };
  }
}

File: kam-admin-module-monitor/src/main/java/io/kamailio/admin/modules/system/service/SystemService.java
package io.kamailio.admin.modules.system.service;

import io.kamailio.admin.common.service.KamailioRpcService;
import org.springframework.stereotype.Service;

import java.util.Map;

@Service
public class SystemService {
    private final KamailioRpcService kamailioRpc;

    public SystemService(KamailioRpcService kamailioRpc) { this.kamailioRpc = kamailioRpc; }

    public Object getCoreInfo() { return kamailioRpc.getCoreInfo().block(); }
    public Object getVersion() { return kamailioRpc.call("core.version").block(); }
    public Object getUptime() { return kamailioRpc.getUptime().block(); }
    public Object getSharedMemory() { return kamailioRpc.call("core.shmmem").block(); }
    public Object getProcessList() { return kamailioRpc.call("core.psx").block(); }
    public Object getConfig(String group, String name) { return kamailioRpc.call("cfg.get", group, name).block(); }
    public void setConfigNow(String group, String name, Object value) {
        if (value instanceof Number) kamailioRpc.call("cfg.set_now_int", group, name, value).block();
        else kamailioRpc.call("cfg.set_now_string", group, name, value.toString()).block();
    }
    public Object listConfig() { return kamailioRpc.call("cfg.list").block(); }
    public Object diffConfig() { return kamailioRpc.call("cfg.diff").block(); }
    public Object getTlsList() { return kamailioRpc.call("tls.list").block(); }
    public Object getTlsInfo() { return kamailioRpc.call("tls.info").block(); }
    public void reloadTls() { kamailioRpc.call("tls.reload").block(); }
    public Object getPikeList() { return kamailioRpc.call("pike.list").block(); }
    public Object getPikeTop(int limit) { return kamailioRpc.call("pike.top", limit).block(); }
    public Object getStatistics(String group) { return kamailioRpc.getStatistics(group).block(); }
    public void resetStatistics(String name) { kamailioRpc.call("stats.reset_statistics", name).block(); }
    public void clearStatistics(String name) { kamailioRpc.call("stats.clear_statistics", name).block(); }
    public Object getModulesList() { return kamailioRpc.call("core.modules").block(); }
    public Object getSystemStatus() {
        Object info = null, uptime = null, memory = null, processes = null;
        try { info = getCoreInfo(); } catch (Exception ignored) {}
        try { uptime = getUptime(); } catch (Exception ignored) {}
        try { memory = getSharedMemory(); } catch (Exception ignored) {}
        try { processes = getProcessList(); } catch (Exception ignored) {}
        int processCount = processes instanceof java.util.List ? ((java.util.List<?>) processes).size() : 0;
        return Map.of("info", info, "uptime", uptime, "memory", memory, "processCount", processCount);
    }
}
